### ISSUE TITLE
Higher order function mishandles type variables and fails to report type errors

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#7618]
```python
from __future__ import annotations

from collections.abc import Callable
from typing import TypeVar

A = TypeVar("A")
B = TypeVar("B")
Int = TypeVar("Int", bound=int)

def lift(fun: Callable[[A], B]) -> Callable[[A], B]:
    raise NotImplementedError

def main(fn: Callable[[Int], float], arg1: int, arg2: str):
    fn(arg1) # Type error
    lift(fn)(arg1) # Unexpected lack of type error
    lift(fn)(arg2) # Type error
```

In the above code, I would expect all 3 lines to produce a type error. Indeed, mypy 1.9.0 does report a type error for all 3 lines. However, pyright 1.1.357 does not report a type error on the middle line. It seems like `lift(fn)` ends up "falling back" to just `int → float` instead of `Int → float` (even though `reveal_type` reports `Int → float`).

**Comment by erictraut** [comment]
Thanks for the bug report. This has the same root cause as several other open issues in the pyright issue tracker.

**Comment by colehaus** [comment]
Yes, sorry, just noticed https://github.com/microsoft/pyright/issues/7542. Feel free to close this one.

**Comment by erictraut** [comment]
Closing as duplicate.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by colehaus** [comment]
This fixed several false positives in my codebase. Thanks!
